### PR TITLE
[airbyte-cdk] - Add `XmlDecoder` component to low code CDK

### PR DESCRIPTION
**Pull request by pnilan**
## What
- Resolves https://github.com/airbytehq/airbyte-internal-issues/issues/10027
- Adds new `XmlDecoder` component for decoding responses returned in XML format, using the `xmltodict` library
- Adds new `PaginationDecoderDecorator` which is used internally to wrap decoders when initializing the specified paginator. This ensures streamed responses will not decode responses for pagination.
- Adds `XmlDecoder` support for `SessionTokenAuthenticator`

## Review guide
1. xml_decoder.py
2. pagination_decoder_decorator.py
3. model_to_component_factory.py
4. declarative_component_schema.yaml
5. tests

## User Impact
- Enables connector devs to specify `XmlDecoder` as the decoder type in a low-code connector manifest for the `SimpleRetriever`, `AsyncRetriever`, and `SessionTokenAuthenticator`

## Out of Scope
- Caching decoded responses for XML
- `XmlDecoder` for file-based sources
- Automatically detecting response data format and applying respective decoder
- Namespace support

## Can this PR be safely reverted and rolled back?

- [X] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by bazarnov** [comment]
@pnilan We have a couple of formatting issues to be resolved:
```
./unit_tests/sources/declarative/decoders/test_xml_decoder.py:21:17: E131 continuation line unaligned for hanging indent
./unit_tests/sources/declarative/decoders/test_xml_decoder.py:23:13: E124 closing bracket does not match visual indentation
```

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/decoders/pagination_decoder_decorator.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import logging
+from typing import Any, Generator, MutableMapping
+
+import requests
+from airbyte_cdk.sources.declarative.decoders import Decoder
+
+logger = logging.getLogger("airbyte")
+
+
+class PaginationDecoderDecorator(Decoder):
```
Comment (bazarnov) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/decoders/pagination_decoder_decorator.py:14]: Is it intentional that we are not using @dataclass for this implementation?

**Review by bazarnov — APPROVED** [review]
Looks good to me! Nice work @pnilan.

**Comment by artem1205** [comment]
I see 1 obvious ~~bug~~ feature about `xml` format we should worry about. Let's assume we have the following `xml` response (RSS like):
```xml
<?xml version="1.0" encoding="utf-8" ?>
<APResponse>
	<resource>locations</resource>
	<action>getlocations</action>
	<request>GetLocationsById</request>
	<result>success</result>
	<errors>
		<error>If there were any errors they would show here.</error>
	</errors>
	<count>1</count>
	<data>
		<location>
			<c_id>49</c_id>
			<location_id>84</location_id>
			<headquarters>N</headquarters>
			<sort_order>2</sort_order>
			<location_name>PA - Pottsville Campus</location_name>
			<headquarters_id>49</headquarters_id>
		</location>
	</data>
</APResponse>
```

with the use of `xmltodict` library we could convert it to nice `json` and then use the `DpathExtractor` to get records from the `/data/**/location` path
image


Unfortunately, there is another style of `xml`, in which some other `attributes` exist in `node` (tag) or even the values, that leads into:
```xml
<crm-item name="citedby-count" type="number">0</crm-item>
<crm-item name="owner-prefix" type="string">10.1353</crm-item>
<crm-item name="citation-id" type="number">25715607</crm-item>
<crm-item name="journal-id" type="number">48965</crm-item>
```

image

Do we need to handle this case ? Should we explicitly declare somewhere (docs?) about `xml`'s format we support?

I guess we should at least:
1. update the class docstring to explain the limitations about this decoder.
2. add tests (with `extractor` to demonstrate real use-cases).

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/decoders/pagination_decoder_decorator.py (hunk 1 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/decoders/pagination_decoder_decorator.py:14]: Nope! good catch. Thanks

**Comment by pnilan** [comment]
@artem1205 

Yes you're 100% correct -- and this should be documented.

We've opted to use the `xmltodict` output format. After some investigation I found it's common to handle node attributes when converting to JSON properties by prefixing them with `@`, which `xmltodict` already does. As for the "text" of a node, `xmltodict` handles by creating a `#text` property. While it may be cumbersome for customers to initially work with data with these headers, we felt it was a straightforward approach right now and would revise based on customer feedback.

For example the XML:

```xml
<response>
    <crm-item name="citedby-count" type="number">0</crm-item>
    <crm-item name="owner-prefix" type="string">10.1353</crm-item>
    <crm-item name="citation-id" type="number">25715607</crm-item>
    <crm-item name="journal-id" type="number">48965</crm-item>
</response>
```
Would be converted to:

```json
{
  "response": {
    "crm-item": [
      {
        "@name": "citedby-count",
        "@type": "number",
        "#text": "0"
      },
      {
        "@name": "owner-prefix",
        "@type": "string",
        "#text": "10.1353"
      },
      {
        "@name": "citation-id",
        "@type": "number",
        "#text": "25715607"
      },
      {
        "@name": "journal-id",
        "@type": "number",
        "#text": "48965"
      }
    ]
  }
}
```

Another note: I currently have namespace support turned off, as it doesn't seem to be *too* common with REST APIs w/ XML. If we're requested namespace support it would be relatively easy to add by default or as an option in the manifest.

I haven't updated the low-code documentation yet so I'll go ahead and do that this morning and make it clear what we do/don't currently support and the expected format of converted XML.

**Pull request merged**